### PR TITLE
SHAP Intergration - Implicit Model Serialization

### PR DESCRIPTION
**Pull request by vivekchettiar**
Added initial implementation of shap log/load with internal inference of underlying model type. Serialization of underlying model using mlflow primitives if possible.

### CODE REVIEW

File: mlflow/shap.py [hunk 1]
```diff
@@ -179,3 +228,234 @@ def predict_function(X) -> pred:
...
+    # saving the underlying model if required
+    if serialize_model_using_mlflow == True:
+        underlying_model_flavor = get_underlying_model_flavor(explainer.model)
+
+        if underlying_model_flavor != _UNKNOWN_MODEL_FLAVOR:
+            explainer.model.save = None
+            underlying_model_path = os.path.join(path, _UNDERLYING_MODEL_SUBPATH)
+        
+        if underlying_model_flavor == 'sklearn':
+            mlflow.sklearn.log_model(explainer.model.model.__self__, underlying_model_path)
```
Comment (eddiestudies) [inline mlflow/shap.py]: .save_model

The goal is not to log_model internally but to save_model internally.

This leaves only 1 model(the explainer) in the system

Comment (eddiestudies) [reply mlflow/shap.py]: If save Model is used, it should be possible to get the underlying conda env from the saved model and use it in a merge with the default one.

It should be a best effort merge with a graceful failure if possible

File: mlflow/shap.py [hunk 2]
```diff
@@ -179,3 +228,234 @@ def predict_function(X) -> pred:
...
+    elif not isinstance(conda_env, dict):
+        with open(conda_env, "r") as f:
+            conda_env = yaml.safe_load(f)
+
+    # merging the conda environment generated by serializing the underlying model
+    if underlying_model_path is not None:
+        underlying_model_conda_path = os.path.join(underlying_model_path, 'conda.yaml')
+        underlying_model_conda_file = open(underlying_model_conda_path, 'r')
+        underlying_model_conda_env = yaml.safe_load(underlying_model_conda_file)
+        conda_env = _merge_environments(conda_env, underlying_model_conda_env)
```
Comment (eddiestudies) [inline mlflow/shap.py:431]: Ha awesome you already did this!

File: mlflow/shap.py (hunk 1 above)
Comment (eddiestudies) [reply mlflow/shap.py]: Nevermind you are already doing the merge

Comment (vivekchettiar) [reply mlflow/shap.py]: Using .save_model()

**Pull request closed without merging**